### PR TITLE
Haskell build fixes

### PR DESCRIPTION
**Pull request by e1528532**
# Purpose

Fix the issues described in #2050 and #2046 

# Checklist

Check relevant points but please do not remove entries.
For docu fixes, spell checking, and similar nothing
needs to be checked.

- [X] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] I added unit tests
- [X] I ran all tests locally and everything went fine
- [X] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)
- [X] release notes are updated (doc/news/_preparation_next_release.md)

@markus2330 please review my pull request

### CODE REVIEW

**Comment by e1528532** [comment]
jenkins build all please

**Comment by ingwinlu** [comment]
I added these commits to #2046 to see if it builds correctly as we do not have any checks in master.

**Comment by e1528532** [comment]
Turns out there is a hidden error:

```
c2hs: C header contains errors:
/usr/include/stdio.h:133: (column 19) [ERROR]  >>> Syntax error !
  The symbol `_close' does not fit here.
```

Of course this doesn't happen on my osx so its even harder to debug or fix it. And i also wonder when this got introduced because i think it used to work at some point.

I've only found this error described in https://github.com/haskell/c2hs/issues/176 where it seems to be resolved with a never version of c2hs, but i am already using the latest version 0.28.5 and most people just try around randomly without a real solution.

Digging further into the issue i noticed that the mentioned header file /usr/include/stdio.h doesn't even exist on my own macOS so i wonder where this comes from. I slowly think it somehow picks the wrong header file and then fails parsing it. Maybe the latest xcode image update caused the issue i'll try reverting it for the haskell image for now.
If that doesn't help either i have no quick idea how to fix it to be honest..

**Comment by e1528532** [comment]
Slowly i think it really picks the wrong stdio.h version. The one that comes with clang on osx should be in
`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/usr/include/stdio.h`
(or something like that)
and not `/usr/include/stdio.h`. I suspect the latter got installed via gcc or something like that and is incompatible with the other header file.
I will try to find a way to point the haskell build to the other directory if that is possible.

**Comment by sanssecours** [comment]
> Slowly i think it really picks the wrong stdio.h version. The one that comes with clang on osx should be in `/Applications/Xcode.app/…/usr/include/stdio.h` (or something like that) and not `/usr/include/stdio.h`.

According to a [StackOverflow question](https://stackoverflow.com/questions/28362994/where-is-my-stdio-h-in-mac) the location `/usr/include/stdio.h` should be correct if you install the command line tools. The file `/usr/include/stdio.h` is also present on my machine and contains exactly the same content as `/Applications/Xcode.app/…/usr/include/stdio.h`:

```sh
diff /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/stdio.h /usr/include/stdio.h
# RET: 0
```

.

**Comment by e1528532** [comment]
I think i have tracked the issue down. Turns out a new version of language-c was released a few days ago that new c2hs installations use by default then. I enforced the previous version now and it looks better:
https://travis-ci.org/e1528532/libelektra/jobs/389146568#L37
I'll report the error to the language-c maintainers and will stick to the previous version for our builds for now.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build libelektra please

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build travis please

**Comment by e1528532** [comment]
jenkins build all please

File: .travis.yml [hunk 1]
```diff
@@ -8,9 +8,13 @@ cache:
 
 # don't cache our own libraries generated in the build
 # so they always get rebuilt properly and to avoid init issues
+# | true with unregister as they might not exist on build errors
 before_cache:
   - cd $HOME/elektra-cabal-sandbox
-  - cabal sandbox hc-pkg unregister spectranslator specelektra libfa libelektra-haskell
+  - cabal sandbox hc-pkg unregister spectranslator | true
```
Comment (sanssecours) [inline .travis.yml]: Typo: `|` → `||`

While piping the output of `cabal sandbox` to `true`, will also work here – by default the shell  ignores the exit status of a command to the left of a pipe – I think you want to run `true` if (and only if) `cabal sandbox` fails. 

I could be wrong though and you always want to ignore the output of `cabal sandbox`. In that case I think the command is a little bit to „clever” 😊.

Comment (e1528532) [reply .travis.yml]: yes i don't really care if it got created during the build unregister it if not it will fail and well then not. so returning true in any case is fine

Comment (sanssecours) [reply .travis.yml]: In that case, please write something like

```bash
cabal sandbox hc-pkg unregister spectranslator > /dev/null; true
```

instead. Otherwise 

- stupid people like me will assume the command is incorrect,
- the command will fail, if some sets the option `pipefail` and `cabal sandbox` fails

.

**Comment by e1528532** [comment]
jenkins build all please

File: .travis.yml [hunk 2]
```diff
@@ -8,13 +8,14 @@ cache:
...
-# | true with unregister as they might not exist on build errors
+# we ignore the result as some or all might not exist on build
+# errors, in that case unregister would fail
 before_cache:
   - cd $HOME/elektra-cabal-sandbox
-  - cabal sandbox hc-pkg unregister spectranslator | true
-  - cabal sandbox hc-pkg unregister specelektra | true
-  - cabal sandbox hc-pkg unregister libfa | true
-  - cabal sandbox hc-pkg unregister libelektra-haskell | true
+  - cabal sandbox hc-pkg unregister spectranslator > /dev/null; true
```
Comment (sanssecours) [inline .travis.yml]: Thank you 💖.

**Comment by e1528532** [comment]
jenkins build libelektra please

**Comment by e1528532** [comment]
Looks like there is still some issue with the cache cleanup that i have to investigate.

However the language-c maintainer reacted quickly to my bug report and already released 0.8.1 that should fix my issue so i don't have to stick to 0.7.2. I will test that as well and revert the changes if it works now (maybe annotating that 0.8 is not working)

**Comment by e1528532** [comment]
@sanssecours i'll cleanup all those commits anyway but since you seem very familiar with travis, do you know if there is a faster way to gather debug infos than triggering travis builds over and over again?

**Comment by sanssecours** [comment]
>  i'll cleanup all those commits anyway but since you seem very familiar with travis, do you know if there is a faster way to gather debug infos than triggering travis builds over and over again?

Not as far as I know. If you really want to the debug this stuff, I would just remove all of the code you do not need. If you just want to check out the environment something like

```yaml
language: cpp

cache:
  directories:
    $HOME/.m2
    $HOME/.cabal
    $HOME/Library/Haskell
    $HOME/elektra-cabal-sandbox

#
# Define the build matrix
#
matrix:
  include:

    # HASKELL: Only build Haskell binding and plugin

    - os: osx
      osx_image: xcode9.3
      compiler: clang
      env:
        - HASKELL=ON

before_cache:
  - ls $HOME/.m2
  - ls $HOME/.cabal
  - ls $HOME/Library/Haskell
  - ls $HOME/elektra-cabal-sandbox

script: 'true'
``` 

should be enough, as far as I can tell. The script above [takes about 30 seconds](https://travis-ci.org/sanssecours/elektra/builds/389744943) to execute.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
@markus2330 same as explained in the regexdispatcher, those few tests in the queue are unlikely to succeed and the failure looks unrelated to my changes as well:

```

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:159: error in test_gpg: failed to install the GPG test key

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:187: error in test_file_crypto_operations: kdb set failed

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:188: error in test_file_crypto_operations: file content did not change during encryption

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:191: error in test_file_crypto_operations: kdb get (pregetstorage) failed

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:195: error in test_file_crypto_operations: kdb get (postgetstorage) failed

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:196: error in test_file_crypto_operations: postgetstorage did not encrypt the file again

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:231: error in test_file_signature_operations: kdb set failed

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:232: error in test_file_signature_operations: file content did not change during encryption

/home/jenkins/workspace/libelektra_PR-2055-S7GIINT4DABLKHJXDO3WPT3BY2GBLP5BFFW3JUZATH2RLN6QOMPQ/src/plugins/fcrypt/testmod_fcrypt.c:235: error in test_file_signature_operations: kdb get failed
```

I think it could be merged safely.

**Comment by markus2330** [comment]
Please rebase. Seems like there was a order in your PRs I did not respect.

File: cmake/Modules/LibAddHaskellPlugin.cmake [hunk 3]
```diff
@@ -40,272 +40,311 @@ macro (add_haskell_plugin target)
...
-										set (CABAL_CUSTOM_SETUP_FILE
-										     "${CMAKE_CURRENT_SOURCE_DIR}/Setup.hs.in")
-										if (NOT EXISTS ${CABAL_CUSTOM_SETUP_FILE})
+		if (NOT BUILD_STATIC AND NOT BUILD_FULL)
+			if (NOT ENABLE_ASAN)
+				if (PLUGINPROCESS_FOUND)
+					if (HASKELL_FOUND)
+						check_binding_included ("haskell" HAVE_HASKELL_BINDING)
+						if (HAVE_HASKELL_BINDING)
+
```
Comment (markus2330) [inline cmake/Modules/LibAddHaskellPlugin.cmake]: please put this code in a macro, it is kind of unreadable that way (too deeply nested)

File: cmake/Modules/LibAddHaskellPlugin.cmake [hunk 4]
```diff
@@ -40,272 +40,311 @@ macro (add_haskell_plugin target)
...
+						remove_plugin (${target} ${HASKELL_NOTFOUND_INFO})
+					endif (HASKELL_FOUND)
+				else (PLUGINPROCESS_FOUND)
+					remove_plugin (${target} "${PLUGINPROCESS_NOTFOUND_INFO}, but required for haskell plugins")
+				endif (PLUGINPROCESS_FOUND)
+			else (NOT ENABLE_ASAN)
+				remove_plugin (${target} "haskell plugins are not compatible with ENABLE_ASAN")
+			endif (NOT ENABLE_ASAN)
+		else (NOT BUILD_STATIC AND NOT BUILD_FULL)
+			remove_plugin (${target} "haskell plugins are not compatible with BUILD_STATIC or BUILD_FULL")
```
Comment (markus2330) [inline cmake/Modules/LibAddHaskellPlugin.cmake]: This is too strict. It will remove your plugin *altogether* if someone has selected these build variants. What you want is to only exclude your plugin from the BUILD_STATIC AND BUILD_FULL builds.

So at the moment, if someone wants to compile Elektra in full and shared, your plugin also gets removed also in shared although it would work there.

File: doc/news/_preparation_next_release.md [hunk 5]
```diff
@@ -214,7 +214,9 @@ Many problems were resolved with the following fixes:
 
 - <<TODO>>
 - <<TODO>>
-- <<TODO>>
+- Haskell plugins and bindings are now correctly excluded when using BUILD_FULL or BUILD_STATIC
+  as this is currently unsupported. Another issue when building haskell plugins with a cached sandbox
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Which issue? These two points are unrelated, you should make two points out of it.

File: src/bindings/haskell/CMakeLists.txt [hunk 6]
```diff
@@ -156,9 +156,9 @@ if (HASKELL_FOUND)
 							       "LD_LIBRARY_PATH=${CMAKE_BINARY_DIR}/lib")
 				endif (ENABLE_KDB_TESTING)
 			endif (BUILD_TESTING)
-		else (NOT BUILD_STATIC)
-			exclude_binding (haskell "BUILD_STATIC is currently not compatible with the haskell bindings")
-		endif (NOT BUILD_STATIC)
+		else (NOT BUILD_STATIC AND NOT BUILD_FULL)
+			exclude_binding (haskell "BUILD_STATIC and BUILD_FULL are currently not compatible with the haskell bindings")
```
Comment (markus2330) [inline src/bindings/haskell/CMakeLists.txt]: Same problem here.

File: src/bindings/haskell/README.md [hunk 7]
```diff
@@ -40,13 +40,17 @@ which contains the current list of haskell packages by invoking `cabal update`,
 is not done automatically. Then first ensure the following programs are available on
 the PATH:
 
-- c2hs (tested with 0.28.2)
+- c2hs (tested with 0.28.2 and 0.28.5)
+  c2hs depends on the language-c library and is usually compiled from scratch when installed
```
Comment (markus2330) [inline src/bindings/haskell/README.md]: Maybe put the error message here so that troubleshooting is easier.

File: cmake/Modules/LibAddHaskellPlugin.cmake (hunk 4 above)
Comment (e1528532) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: is it possible to only exclude it from certain builds? should i not remove it upon build/static and build/full but just don't call any functionality in that case? (e.g. a return on top)

Comment (markus2330) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: You need to add a "ONLY_SHARED" feature in cmake/Modules/LibAddPlugin.cmake and in this case the static/full should not be added. (In a quick look it looks like only an if (NOT ONLY_SHARED) around the last two lines is needed.)

Comment (e1528532) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: should it emit some message in that case?

Comment (markus2330) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: You need to handle the case if the plugin should be included but BUILD_SHARED is not active (maybe exclude the plugin then?). And you could update the "Include Plugin regexdispatcher" -> "Include Plugin regexdispatcher for shared builds" or similar.

Adding further new lines are most likely overlooked and rather increase confusion.

File: src/bindings/haskell/README.md (hunk 7 above)
Comment (e1528532) [reply src/bindings/haskell/README.md]: I don't think thats helpfull as you need to compile with -v2 to see the full error message that is not the default. But i think it won't really be that much of an issue due to the language-c update on hackage.

**Comment by e1528532** [comment]
@markus2330 added ONLY_SHARED keyword to add_plugin and deal with all issues that arose.

I had to exclude such plugins from the c file reflecting the exported_symbols otherwise the full and static builds fail. This is even handled dynamically. To me it looks like the shared version still works fine that way. You might want to review it again as the last commit changed quite alot on that matter.

File: src/libs/elektra/exportsymbols.c.in [hunk 8]
```diff
@@ -9,6 +9,22 @@
 #include <stdio.h>
 #include <string.h>
 
+// clang-format off
+static const char * excludedPlugins[] = { @SHARED_ONLY_PLUGINS_ARRAY@ };
```
Comment (markus2330) [inline src/libs/elektra/exportsymbols.c.in]: I am afraid that won't work for cross-compilation. Why not simply remove the "SHARED_ONLY_PLUGINS" on CMake level (src/libs/elektra/CMakeLists.txt line 25 and 29) and simply not pass them as arguments to exportsymbols.

File: cmake/Modules/LibAddPlugin.cmake [hunk 9]
```diff
@@ -394,7 +400,20 @@ function (add_plugin PLUGIN_SHORT_NAME)
...
+			       "it can only be built if BUILD_SHARED is enabled, both BUILD_FULL or BUILD_STATIC may be enabled as well")
+		return ()
+	endif (ARG_ONLY_SHARED AND NOT BUILD_SHARED)
+
+	set (STATUS_MESSAGE "Include Plugin ${PLUGIN_SHORT_NAME}")
+	if (ARG_ONLY_SHARED)
+
+		# also add it to the list of ONLY_SHARED plugins for exportsymbols.c configuration
+		set (STATUS_MESSAGE "${STATUS_MESSAGE} for shared builds")
+		set_property (GLOBAL APPEND PROPERTY SHARED_ONLY_PLUGINS "${PLUGIN_SHORT_NAME}")
```
Comment (markus2330) [inline cmake/Modules/LibAddPlugin.cmake]: Please document what SHARED_ONLY_PLUGINS are for. (For example, see cmake/ElektraCache.cmake line 16)

File: cmake/Modules/LibAddHaskellPlugin.cmake [hunk 10]
```diff
@@ -375,3 +130,194 @@ macro (add_haskell_plugin target)
...
+							    COMMAND ${CABAL_EXECUTABLE} build -v0 || ( ${CABAL_EXECUTABLE} configure
+								    ${CABAL_OPTS} -v0 && ${CABAL_EXECUTABLE} build -v0 )
+							    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+							    DEPENDS c2hs_haskell
+								    ${PLUGIN_SOURCE_FILES}
+								    ${HASKELL_ADD_SOURCES_TARGET}
+							    APPEND)
+					add_custom_target (${target} ALL DEPENDS ${PLUGIN_HASKELL_NAME})
+
+					if (BUILD_SHARED OR BUILD_FULL)
```
[the remark this answers is not in the document]
Comment (e1528532) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: those are some leftovers when i initially intended to support it, i can remove those

File: cmake/Modules/LibAddBinding.cmake [hunk 11]
```diff
@@ -87,6 +87,12 @@ endfunction (check_binding_included)
 #    add_binding ("anynameyouwant")
 # ~~~
 function (add_binding BINDING_NAME)
+	cmake_parse_arguments (ARG
+			       "ONLY_SHARED" # optional keywords
```
[the remark this answers is not in the document]
Comment (e1528532) [reply cmake/Modules/LibAddBinding.cmake]: yes since bindings don't seem to be bundled with FULL or STATIC so for those its just an info. Or should i remove it again since it doesn't really seem to matter?

[the remark this answers is not in the document]
Comment (markus2330) [reply cmake/Modules/LibAddBinding.cmake]: No it is okay. But aren't all bindings ONLY_SHARED currently?

[the remark this answers is not in the document]
Comment (e1528532) [reply cmake/Modules/LibAddBinding.cmake]: i don't know how its the case for the other bindings but for the haskell ones they only work with the shared version yes, thats why i only added it to them.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
@markus2330 next round ;) but good hints the build code gets tidied up a lot it was really a bit messy already.

**Comment by e1528532** [comment]
just saw that i've overlooked removing a HEAD from the release notes, but that little change doesn't really justify rerunning all tests the results of the previous commit still hold its just a docu change.

**Comment by markus2330** [comment]
Thank you, great job!
